### PR TITLE
[lemmas] Reify info for implicits, universe decls, and rec theorems.

### PR DESCRIPTION
**Pull request by ejgallego**
Key information about an interactive lemma proofs was stored as a
closure on an ad-hoc hook, then later made available to the hook
closing actions.

Instead, we put this information in the lemma state and incorporate
these declarations into the normal save path. This includes breaking down the `kind`
triple as their components do belong to different layers.

We also do a small refactoring on `comFixpoint`, and refactor proof information into something more principled, including the removal of duplicate proof information among layers.

After this PR the situation is that we have 3 proof "layers":
- `Proof.t`: contains basic proof objects and goal state, plus some other state such as `polymorphic` and `name` which is used by the abstract tactics
- `Proof_global.t`: contains a `Proof.t` plus the information needed to perform the `close_proof` action, that is to say, to generate an entry suitable for abstract of for workers [par?]
- `Lemma.t` contains a `Proof_global.t` plus the required information to save a constant, including hooks, termination-specific information such as locality or obligation data, etc...

Overlays:
- https://github.com/LPCIC/coq-elpi/pull/66
- https://github.com/mattam82/Coq-Equations/pull/222
- https://github.com/Mtac2/Mtac2/pull/215
- https://github.com/coq-community/paramcoq/pull/35
- https://github.com/QuickChick/QuickChick/pull/167

### CODE REVIEW

**Comment by ejgallego** [comment]
This is of interest to #10297 , even if not a solution [yet] to the problem, it does put critical code here in a direct style which should make the fix easier.

**Comment by ejgallego** [comment]
@SkySkimmer I've pushed part of the proof info refactoring here; there are some bits more, but I need to merge some stuff before rebasing my 3-months branches.

**Comment by ejgallego** [comment]
Ok folks, this is ready and should be the last large PR in this area for a while [next big stuff will be likely trying to move global up?]

This PR does 3 things:
- removes the key information about "recthms" from the hook and places it in direct style; this allows clients to delegate multi-contant proofs correctly if they want, and makes the code in lemmas much clearer, revealing some surprising code paths.
- removes duplicate information on some layers, for example `poly` was stored twice, and tries to organize all this information better.
- moves data from `Decl_kinds` to better places (cc: @maximedenes ) , in particular it breaks the "kind" triple into principled components.

The PR should introduce no functional change.

File: interp/declare.mli [hunk 1]
```diff
@@ -24,7 +24,7 @@ open Decl_kinds
 
 type section_variable_entry =
   | SectionLocalDef of Evd.side_effects definition_entry
-  | SectionLocalAssum of types Univ.in_universe_context_set * polymorphic * bool (** Implicit status *)
+  | SectionLocalAssum of types Univ.in_universe_context_set * bool (* polymorphic *) * bool (* Implicit status *)
```
Comment (SkySkimmer) [inline interp/declare.mli]: Maybe use a local record here?

File: library/decls.ml [hunk 2]
```diff
@@ -18,7 +18,7 @@ open Libnames
 (** Datas associated to section variables and local definitions *)
 
 type variable_data =
-  DirPath.t * bool (* opacity *) * Univ.ContextSet.t * polymorphic * logical_kind
+  DirPath.t * bool (* opacity *) * Univ.ContextSet.t * bool (* poly *) * logical_kind
```
Comment (SkySkimmer) [inline library/decls.ml]: Maybe make this a record?

File: library/lib.ml [hunk 3]
```diff
@@ -411,8 +411,8 @@ type abstr_info = {
 type abstr_list = abstr_info Names.Cmap.t * abstr_info Names.Mindmap.t
 
 type secentry =
-  | Variable of (Names.Id.t * Decl_kinds.binding_kind *
-		   Decl_kinds.polymorphic * Univ.ContextSet.t)
+  | Variable of (Names.Id.t * Decl_kinds.binding_kind * bool * Univ.ContextSet.t)
+  (** (name, kind, poly, univs) *)
```
Comment (SkySkimmer) [inline library/lib.ml]: Record?

**Review by SkySkimmer — APPROVED** [review]

File: interp/declare.mli (hunk 1 above)
Comment (ejgallego) [reply interp/declare.mli]: Sure, can you suggest some names? [Feel free just to push to the branch if you think is quicker]

File: library/decls.ml (hunk 2 above)
Comment (ejgallego) [reply library/decls.ml]: See #10417 , in fact I'd like to get rid of `Decls` and `Decl_kinds` altogether as it is only used for Dumpglob [I mean, moving this info to Dumpglob], so in this case I'd rather see if I can do a PR removing that, running CI on it now.

File: library/lib.ml (hunk 3 above)
Comment (ejgallego) [reply library/lib.ml]: Sure, feel free to push or to suggest a naming scheme.

File: interp/declare.mli (hunk 1 above)
Comment (SkySkimmer) [reply interp/declare.mli]: Pushed

**Comment by gares** [comment]
> After this PR the situation is that we have 3 proof "layers":
> 
>     * `Proof.t`: contains basic proof objects and goal state, plus some other state such as `polymorphic` and `name` which is used by the abstract tactics
> 
>     * `Proof_global.t`: contains a `Proof.t` plus the information needed to perform the `close_proof` action, that is to say, to generate an entry suitable for abstract of for workers [par?]
> 
>     * `Lemma.t` contains a `Proof_global.t` plus the required information to save a constant, including hooks, termination-specific information such as locality or obligation data, etc...

This clarification looks good, but the names of the modules are not matching their actual content.

May I suggest `[Ongoing]Proof.t`, `ClosedProof.t` and `SavedProof.t` respectively.

In particular `Proof_global.t` is misleading, since it is not "global", not in the old sense (as a "global" state) nor in the new sense (since it becomes "global" once it is saved, that is the third layer, not the second).

**Comment by ejgallego** [comment]
> May I suggest `[Ongoing]Proof.t`, `ClosedProof.t` and `SavedProof.t` respectively.

I dunno, for me `Proof` and `Lemmas` are OK, I'm not a fan of these "verbal" names, so not sure worth the noise in history that such a renaming would bring.

`Proof_global` is indeed a bad name but the module seems to be leaning towards disappearing so that's why I didn't bother too much; ideally we would implement the plan in #9146 and the module should go away.

Anyways the status-quo described above was already the case before this PR, I just documented it and removed a few duplications; so IMHO we should do the renaming in a separate PR.

**Comment by ejgallego** [comment]
> `ClosedProof.t` and `SavedProof.t` respectively.

Indeed actually that'd be `CloseableProof` and `ProofWithConstant`, in both cases the proof is not closed or saved yet so I think the tense would be misleading.

**Comment by ejgallego** [comment]
Bulk of this PR was written 3/2 months ago, I don't have more changes planned so maybe it is the time to get an assignee, @ppedrot seems like the natural one based on component maintainership.

The changes themselves are straightforward but indeed review may bring some fine-tuning.

This needs coordination with #10406 , I am fine with any order.

**Comment by ejgallego** [comment]
PR and overlays rebased after two conflicts, IMHO this is a good moment to try to merge this @ppedrot

**Review by ppedrot — APPROVED** [review]
I haven't seen anything suspicious, and the commits make sense. I cannot say I am an expert in this part of  Coq, but this looks reasonable enough for me to give my formal approval.

**Comment by ppedrot** [comment]
I'm waiting for @gares comments, he's the main one who kneaded this spaghetti code, so it's fair to let him act first.

**Comment by ejgallego** [comment]
Cool, thanks @ppedrot ; I think that a large part of the code this PR touches is in fact around 10 year old and due to @herbelin , some other parts are likely due to the new proof engine.

**Comment by gares** [comment]
The changes are ok.

The ~poly label is horrible, but the original sin is in using "poly" for universe polymorphic (since "poly" morphic is a standard term for another thing) so I guess I cannot complain too much. Still, perpetuating the bad choice was avoidable, maybe next time ask first what people think about the names of function arguments before putting the existing name also in the types (and propagate it all over the place).

**Comment by ejgallego** [comment]
@gares I think this PR is a good opportunity to change the label name, what name would you choose?

**Comment by gares** [comment]
Even upoly would be better to me. univpoly would be best, but it is also longer.

Note that I'm not asking you to do the renaming, and actually doing it is going to inflate the pr a lot. Also other should comment on the preferred name before you spend time doing any renaming.

**Comment by ejgallego** [comment]
I guess `upoly` looks good; I liked polymorphic but most code was using `poly` so I defaulted to that. I dunno, indeed no name seems very great.

**Comment by ejgallego** [comment]
So what do you think @ppedrot ? I don't mind doing the renaming but IMHO the position even for Enrico doesn't seem clear [there is a cost indeed] If I rename anyways that'll be a commit of top of this, given that quite a bunch of PRs depend on this I would not block the merge on the renaming alone, but OMMV.

**Comment by ppedrot** [comment]
I personally prefer the `poly` name because I grew accustomed to it, even if it's a bad name originally. If I were the only one there, I would merge as it. @gares?

**Comment by gares** [comment]
Please merge as is.

**Comment by ppedrot** [comment]
@gares can you formally approve so that it appears in the commit message? Thanks.

**Review by gares — APPROVED** [review]

**Comment by mattam82** [comment]
+1 for `univpoly` everywhere, that’s more to the point, but of course I agree that can be done separately.

**Comment by ejgallego** [comment]
Thanks folks!

**Pull request merged**